src/protocol.rs

```diff
@@ -1,42 +1,541 @@
+    ( $name:ident ) => {};
+    ( $name:ident , $( $rest:tt )+ ) => {
+        derive_for_empty_segment!($name : $($rest)*);
+    };
+
+    ( $name:ident : Display $( $rest:tt )* ) => {
+        impl fmt::Display for $name {
+            fn fmt(&self, _: &mut fmt::Formatter) -> fmt::Result {
+                Ok(())
+            }
+        }
+
+        derive_for_empty_segment!($name $($rest)*);
+    };
+
+    ( $name:ident : FromStr $( $rest:tt )* ) => {
+        impl FromStr for $name {
+            type Err = Error;
+
+            fn from_str(_: &str) -> Result<Self> {
+                Ok( $name {} )
+            }
+        }
+
+        derive_for_empty_segment!($name $($rest)*);
+    };
+
+    ( $name:ident : AddressSegment < $proto:ident > $( $rest:tt )* ) => {
+        impl AddressSegment for $name {
+            fn protocol(&self) -> Protocol {
+                Protocol::$proto
+            }
+
+            fn from_stream(_: &mut io::Read) -> Result<Self> {
+                Ok($name {})
+            }
+
+            fn to_stream(&self, _: &mut io::Write) -> io::Result<()> {
+                Ok(())
+            }
+        }
+
+        derive_for_empty_segment!($name $($rest)*);
+    };
+}
+
+
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct IP4Segment(pub Ipv4Addr);
+
+derive_for_wrapping_segment!(IP4Segment: Display, FromStr, From<Ipv4Addr>);
+impl AddressSegment for IP4Segment {
+    const STREAM_LENGTH: usize = 4;
+
+    fn protocol(&self) -> Protocol {
+        Protocol::IP4
+    }
+
+    fn from_stream(stream: &mut io::Read) -> Result<Self> {
+        let mut buf = [0u8; 4];
+        stream.read_exact(&mut buf)?;
+
+        Ok(IP4Segment(Ipv4Addr::new(buf[0], buf[1], buf[2], buf[3])))
+    }
+
+    fn to_stream(&self, stream: &mut io::Write) -> io::Result<()> {
+        stream.write_all(&self.0.octets())
+    }
+}
+
+
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct TCPSegment(pub u16);
+
+derive_for_wrapping_segment!(TCPSegment: Display, FromStr, From<u16>);
+impl AddressSegment for TCPSegment {
+    const STREAM_LENGTH: usize = 2;
+
+    fn protocol(&self) -> Protocol {
+        Protocol::TCP
+    }
+
+    fn from_stream(stream: &mut io::Read) -> Result<Self> {
+        Ok(TCPSegment(stream.read_u16::<BigEndian>()?))
+    }
+
+    fn to_stream(&self, stream: &mut io::Write) -> io::Result<()> {
+        Ok(stream.write_u16::<BigEndian>(self.0)?)
+    }
+}
+
+
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct UDPSegment(pub u16);
+
+derive_for_wrapping_segment!(UDPSegment: Display, FromStr, From<u16>);
+impl AddressSegment for UDPSegment {
+    const STREAM_LENGTH: usize = 2;
+
+    fn protocol(&self) -> Protocol {
+        Protocol::UDP
+    }
+
+    fn from_stream(stream: &mut io::Read) -> Result<Self> {
+        Ok(UDPSegment(stream.read_u16::<BigEndian>()?))
+    }
+
+    fn to_stream(&self, stream: &mut io::Write) -> io::Result<()> {
+        Ok(stream.write_u16::<BigEndian>(self.0)?)
```

Comment:
result types are the same. it is enough to write

```rust
stream.write_u16::<BigEndian>(self.0)
```

Reply:
Same